sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,919 @@
+          throw new IOException(HadoopInputFormatIOContants.CREATEREADER_NULLSPLIT_ERROR_MSG);
+        }
+      } else {
+        return new HadoopInputFormatReader<Object>(this, simpleFuncForKeyTranslation,
+            simpleFuncForValueTranslation, serializableSplit.getSplit());
+      }
+    }
+
+    /** BoundedReader for HadoopInputFormatSource */
+    class HadoopInputFormatReader<T extends Object> extends BoundedSource.BoundedReader<KV<K, V>> {
+
+      private final HadoopInputFormatBoundedSource<K, V> source;
+      private final InputSplit split;
+      private final SimpleFunction<T, K> simpleFuncForKeyTranslation;
+      private final SimpleFunction<T, V> simpleFuncForValueTranslation;
+
+      public HadoopInputFormatReader(HadoopInputFormatBoundedSource<K, V> source,
+          SimpleFunction keyTranslation, SimpleFunction valueTranslation, InputSplit split) {
+        this.source = source;
+        this.split = split;
+        this.simpleFuncForKeyTranslation = keyTranslation;
+        this.simpleFuncForValueTranslation = valueTranslation;
+      }
+
+      @Override
+      public HadoopInputFormatBoundedSource<K, V> getCurrentSource() {
+        return source;
+      }
+
+      private RecordReader<T, T> currentReader;
+      private KV<K, V> currentRecord;
+      private volatile boolean doneReading = false;
+      private long recordsReturned = 0L;
+
+      @Override
+      public boolean start() throws IOException {
+        try {
+          TaskAttemptContextImpl attemptContext = new TaskAttemptContextImpl(
+              source.getConfiguration().getHadoopConfiguration(), new TaskAttemptID());
+          Job job = Job.getInstance(source.getConfiguration().getHadoopConfiguration());
+          InputFormat<?, ?> inputFormatObj =
+              (InputFormat<?, ?>) job.getInputFormatClass().newInstance();
+          currentReader =
+              (RecordReader<T, T>) inputFormatObj.createRecordReader(split, attemptContext);
+          if (currentReader != null) {
+            currentReader.initialize(split, attemptContext);
+            if (currentReader.nextKeyValue()) {
+              currentRecord = nextPair();
+              return true;
+            }
+          } else {
+            throw new IOException(String.format(
+                HadoopInputFormatIOContants.NULL_CREATE_RECORDREADER_ERROR_MSG, inputFormatObj));
+          }
+        } catch (Exception e) {
+          throw new IOException(e);
+        }
+        currentReader = null;
+        currentRecord = null;
+        doneReading = true;
+        return false;
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        try {
+          if (currentReader != null && currentReader.nextKeyValue()) {
+            currentRecord = nextPair();
+            return true;
+          }
+          currentRecord = null;
+          doneReading = true;
+          return false; // no more records to read
+        } catch (Exception e) {
```

Comment:
genuine question: Is this good java practice? Should we instead just catch the InterruptionException and then throw that as an IOException?